### PR TITLE
HPCC-21784 Add option to ForeignLogicalFileName to suppress cluster prefix

### PR DESCRIPTION
**Pull request by jakesmith**
…refix

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by jakesmith** [comment]
@dcamper - FYI

@AttilaVamos - please review

**Review by AttilaVamos — APPROVED** [review]
@jakesmith It seems good.

**Comment by jakesmith** [comment]
@richardkchapman - please merge.

File: docs/EN_US/ECLStandardLibraryReference/SLR-Mods/ForeignLogicalFileName.xml [hunk 1]
```diff
@@ -46,6 +47,14 @@
           the default is FALSE.</entry>
         </row>
 
+        <row>
+          <entry><emphasis>omitClusterPreifx</emphasis></entry>
+
+          <entry>Optional. A boolean TRUE/FALSE to indicate whether the target
+          clusters prefix should automatically be added if 'filename' is a
```
Comment (richardkchapman) [inline docs/EN_US/ECLStandardLibraryReference/SLR-Mods/ForeignLogicalFileName.xml]:
```suggestion
          cluster's prefix should automatically be added if 'filename' is a
```

File: plugins/fileservices/fileservices.cpp [hunk 2]
```diff
@@ -1611,10 +1611,14 @@ FILESERVICES_API void FILESERVICES_CALL fslFinishSuperFileTransaction(ICodeConte
...
-FILESERVICES_API char *  FILESERVICES_CALL fsForeignLogicalFileName(ICodeContext *ctx, const char *_lfn,const char *foreigndali,bool abspath)
+static char * implementForeignLogicalFileName(ICodeContext *ctx, const char *_lfn, const char *foreigndali, bool abspath, bool omitClusterPrefix)
 {
     StringBuffer lfns;
-    constructLogicalName(ctx, _lfn, lfns);
+
+    Owned<IConstWorkUnit> wu;
+    if (!omitClusterPrefix && _lfn&&(*_lfn != '~'))
+        wu.setown(getWorkunit(ctx));
+    constructLogicalName(wu, _lfn, lfns);
```
Comment (richardkchapman) [inline plugins/fileservices/fileservices.cpp]: Can wu end up NULL (in Roxie, for example)?

**Comment by richardkchapman** [comment]
@jakesmith Couple of questions

File: plugins/fileservices/fileservices.cpp (hunk 2 above)
Comment (jakesmith) [reply plugins/fileservices/fileservices.cpp]: it's okay for wu to be null, code in constructLogicalName(IConstWorkUnit *,..
will check for null.

NB: in this regard, the logic is the same as before this PR, i.e. it used to unconditionally get a 'wu' (if not absolute prefix), and then call on as it does now. 'wu' could still be null.

**Comment by jakesmith** [comment]
@richardkchapman - fixed doc. issue + answered other question.

**Comment by richardkchapman** [comment]
@jakesmith A test in regression suite would be good

**Comment by richardkchapman** [comment]
I think this is going to cause compatibility issues if a query compiled with older version tries to link to new libfileservices or vice versa (e.g. deployed roxie queries). May need to leave the old function present and reference a new one

**Comment by jakesmith** [comment]
@richardkchapman - the old exported symbol 'fsForeignLogicalFileName' is still there/the same.
With a new symbol introduced 'fsForeignLogicalFileName_v2' for the variance and the EclDefinition pointing to it. This is the technique used before afaics, to avoid this problem.
So I think older compiles linking to this newer version of fileservices should be ok?

**Comment by richardkchapman** [comment]
Sorry, missed that.

@ghalliday Canyou sanity check

**Comment by ghalliday** [comment]
@richardkchapman the changes looks good.  I agree a regression test would be useful.

**Comment by richardkchapman** [comment]
@jakesmith Did you want to add a test?

**Comment by jakesmith** [comment]
I'll have a look, not sure how to off hand, because it would generate a different prefix on each engine it ran on, and I'm not sure if there's a way to ask for the cluster prefix (but I could add a new getClusterPrefix() method too).

**Comment by jakesmith** [comment]
@ghalliday - have extended the fileservice.ecl regression test (squashed in).

File: testing/regress/ecl/fileservice.ecl [hunk 3]
```diff
@@ -47,4 +53,8 @@ SEQUENTIAL(
   File.RenameLogicalFile(prefix + 'renametest.d00', prefix + 'afterrename1.d00', true),
   output(compareDatasets(ds, ds2)),
   File.DeleteLogicalFile(prefix + 'afterrename1.d00'),
+  RemovePrefix(File.ForeignLogicalFileName('somescope::somefilename', '192.168.168.168', true), 3);
```
Comment (richardkchapman) [inline testing/regress/ecl/fileservice.ecl]: That ip looks like it won't work on most environments?

**Comment by richardkchapman** [comment]
@jakesmith One question re the test case

File: testing/regress/ecl/fileservice.ecl (hunk 3 above)
Comment (jakesmith) [reply testing/regress/ecl/fileservice.ecl]: It is a dummy IP. The function only uses it to construct a logical filename, e.g. ~foreign::192.168.168.168::blah

**Comment by jakesmith** [comment]
@richardkchapman - see reply above.

**Comment by richardkchapman** [comment]
@ghalliday

**Pull request merged**